src/compiler/scala/tools/nsc/backend/jvm/PerRunInit.scala

```diff
@@ -0,0 +1,18 @@
+package scala.tools.nsc.backend.jvm
+
+import scala.collection.mutable.ListBuffer
+
+/**
+ * Utility for backend components that have state that needs to be re-initialized at every compiler
+ * run, for example state that depends on compiler settings of frontend types (Symbols, Types).
+ *
+ * The trait provides an `initialize` method that runs all initializers added through `perRunLazy`.
+ */
+trait PerRunInit {
+  private val inits = ListBuffer.empty[() => Unit]
+
+  def perRunInit(init: => Unit): Unit = inits += (() => init)
```

Comment:
Is it an error to call this after `initialize` has been called? Can we add an assertion?